core/src/main/java/io/questdb/griffin/engine/functions/window/LagDoubleFunctionFactory.java

```diff
@@ -71,253 +65,82 @@ public Function newInstance(
             CairoConfiguration configuration,
             SqlExecutionContext sqlExecutionContext
     ) throws SqlException {
-        final WindowContext windowContext = sqlExecutionContext.getWindowContext();
-        if (windowContext.isEmpty()) {
-            throw SqlException.emptyWindowContext(position);
-        }
-
-        if (args.size() > 3) {
-            throw SqlException.$(argPositions.getQuick(3), "too many arguments");
-        }
-
-        long offset = 1;
-        if (args.size() >= 2) {
-            final Function offsetFunc = args.getQuick(1);
-            if (!offsetFunc.isConstant() && !offsetFunc.isRuntimeConstant()) {
-                throw SqlException.$(argPositions.getQuick(1), "offset must be a constant");
-            }
-
-            offset = offsetFunc.getLong(null);
-            if (offset < 0) {
-                throw SqlException.$(argPositions.getQuick(1), "offset must be a positive integer");
-            }
-        }
-
-        Function defaultValue = null;
-        if (args.size() == 3) {
-            defaultValue = args.getQuick(2);
-            if (!(defaultValue instanceof DoubleFunction)) {
-                throw SqlException.$(argPositions.getQuick(2), "default value must be a double");
-            }
-
-            if (defaultValue instanceof WindowFunction) {
-                throw SqlException.$(argPositions.getQuick(2), "default value can not be a window function");
-            }
-        }
-
-        if (offset == 0) {
-            return new LeadLagValueCurrentRow(windowContext.getPartitionByRecord(), args.get(0), NAME, windowContext.isIgnoreNulls());
-        }
-
-        if (windowContext.getPartitionByRecord() != null) {
-            Map map = MapFactory.createUnorderedMap(
-                    configuration,
-                    windowContext.getPartitionByKeyTypes(),
-                    LAG_COLUMN_TYPES
-            );
-            MemoryARW mem = Vm.getCARWInstance(configuration.getSqlWindowStorePageSize(),
-                    configuration.getSqlWindowStoreMaxPages(), MemoryTag.NATIVE_CIRCULAR_BUFFER
-            );
-
-            return new LagOverPartitionFunction(
-                    map,
-                    windowContext.getPartitionByRecord(),
-                    windowContext.getPartitionBySink(),
-                    mem,
-                    args.get(0),
-                    windowContext.isIgnoreNulls(),
-                    defaultValue,
-                    offset
-            );
-        }
-
-        MemoryARW mem = Vm.getCARWInstance(
-                configuration.getSqlWindowStorePageSize(),
-                configuration.getSqlWindowStoreMaxPages(),
-                MemoryTag.NATIVE_CIRCULAR_BUFFER
+        return LeadLagWindowFunctionFactoryHelper.newInstance(
+                position,
+                args,
+                argPositions,
+                configuration,
+                sqlExecutionContext,
+                (defaultValue) -> {
```

Comment:
this can be replaced with more consistent:

```java
                    if (!ColumnType.isAssignableFrom(defaultValue.getType(), ColumnType.DOUBLE)) {
                        throw SqlException.$(argPositions.getQuick(2), "default value must be can cast to double");
                    }

```

Reply:
fix it